pkg/util/ppcpuusage/cpuusages.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ppcpuusage
+
+import (
+	"sync"
+	"time"
+)
+
+// CPUUsages records tidb/tikv cpu usages
+type CPUUsages struct {
+	TidbCPUTime time.Duration
+	TikvCPUTime time.Duration
+}
+
+// SQLCPUUsages is used to record sqlID and its cpu usages
+type SQLCPUUsages struct {
+	sync.Mutex
+	sqlID     uint64
+	cpuUsages CPUUsages
+}
+
+// Reset resets all cpu times to 0
+func (c *CPUUsages) Reset() {
+	c.TikvCPUTime = 0
+	c.TidbCPUTime = 0
+}
+
+// SetCPUUsages sets cpu usages value
+func (c *SQLCPUUsages) SetCPUUsages(usage CPUUsages) {
+	c.Lock()
+	defer c.Unlock()
+	c.cpuUsages = usage
+}
+
+// MergeTidbCPUTime merges tidbCPU time into self when sqlID matches
+// Checks sqlID here, because tidb cpu time can only be collected by profiler now, and updated in concurrent goroutines
+func (c *SQLCPUUsages) MergeTidbCPUTime(sqlID uint64, d time.Duration) {
+	c.Lock()
+	defer c.Unlock()
+	if c.sqlID == sqlID {
+		c.cpuUsages.TidbCPUTime += d
+	}
```

Comment:
I'm afraid this will miss a lot of data. Here is an example:

- In 1 second, conn-1 executed 5 SQL statements;
- In the cpu profile of 1 second, it collected following infos:
    - conn1_sql1: 200ms
    - conn1_sql2: 200ms
    - conn1_sql3: 200ms
    - conn1_sql4: 200ms
    - conn1_sql5: 200ms
 
 When call `MergeTidbCPUTime`, the `sqlID` must be 5 or over, then the SQL CPU of conn1_sql1, conn1_sql2, conn1_sql3, conn1_sql4 will be missed.

Reply:
Yes, you're right, and this is expected, the feature is designed to collect sampling data, not full data. And for future memory table persistency, it should persistent data right after the Profiling.